files/migration_steps.py

```diff
@@ -0,0 +1,123 @@
+"""
+General purpose migration steps and utilities
+"""
+
+import shutil
+from abc import ABC, abstractmethod
+from pathlib import Path
+
+
+class MigrationStep(ABC):
+    """
+    Abstract base class to specify the API for migration steps
+    """
+
+    def validate(self):
+        return True
+
+    def snapshot(self, tmpdir):
+        pass
+
+    @abstractmethod
+    def run(self):  # pragma: no cover
+        pass
+
+    def rollback(self, tmpdir):
+        pass
+
+    def cleanup(self, tmpdir):
+        pass
+
+    def __repr__(self):
+        return f"<{self.__class__.__name__}: {self.__dict__}>"
+
+    def __str__(self):
+        return self.__repr__()
+
+
+def path_remove(path):
+    if path.is_dir():
+        shutil.rmtree(path)
+    else:
+        path.unlink()
+
+
+def path_validate(path, check_exists):
+    # None: no check
+    # True: check if it exists
+    # False: check if it doesn't exist
+    if check_exists is not None:
+        return path.exists() == check_exists
+    return True
+
+
+def path_snapshot(path, tmpdir):
+    if path.is_dir():
+        shutil.copytree(path, tmpdir / path.name)
+    else:
+        shutil.copy2(path, tmpdir)
+
+
+def path_rollback(path, tmpdir):
+    shutil.move(f"{tmpdir}/{path.name}", path)
+
+
+class Absent(MigrationStep):
```

Comment:
I added more documentation for this, I hope that makes the reason for this naming a bit clearer. I changed the name because I found naming it `Absent` to be more correct than `Remove`, because in some cases nothing would be removed.

Reply:
Ack!